### PR TITLE
HIVE-25057: Implement rollback for hive to iceberg migration

### PR DESCRIPTION
**Pull request by lcspinter**
### What changes were proposed in this pull request?

In case of an issue during the table migration this logic is followed:
- drop altered table if it exists but keep the data
- recreate the original table 
- call `msck repair` on new table

Work performed:
- Enhance `HiveMetaHook` with rollback method for alter operation and provide implementation in `HiveIcebergMetaHook`
- add drop/create/msck repair logic to `AbstractAlterTableOperation`
- the need for rollback is signalled through the `EnvironmentContext` properties. The `HiveMetaHook#INITIALIZE_ROLLBACK_ALTER` is set in `HiveIcebergMetaHook#rollbackAlterTable` and evaluated in `AbstractAlterTableOperation`
- Introduced a new `preDropTable` method to `HiveMetaHook` which accepts the `deleteData` parameter in order to retain data files while deleting iceberg tables.
- covered rollback with unit tests.

### Why are the changes needed?
In case of an error during the migration of a hive table to iceberg the original table must be restored.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Manual test and unit tests.

### CODE REVIEW

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 1]
```diff
@@ -254,6 +262,32 @@ public void commitAlterTable(org.apache.hadoop.hive.metastore.api.Table hmsTable
...
+      // iceberg table was not yet created, no need to delete the metadata dir separately
+      return;
+    }
+    // we want to keep the data files but get rid of the metadata directory
+    hmsTable.getParameters().put(InputFormatConfig.EXTERNAL_TABLE_PURGE, "FALSE");
+    String metadataLocation = ((BaseTable) this.icebergTable).operations().current().metadataFileLocation();
+    try {
+      Path path = new Path(metadataLocation).getParent();
+      FileSystem fileSystem = FileSystem.get(path.toUri(), conf);
+      if (fileSystem.exists(path)) {
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:282]: I think we do not need exists. Just try / catch around it

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 2]
```diff
@@ -254,6 +262,32 @@ public void commitAlterTable(org.apache.hadoop.hive.metastore.api.Table hmsTable
...
+      throws MetaException {
+    context.getProperties().put(INITIALIZE_ROLLBACK_ALTER, "true");
+    this.catalogProperties = getCatalogProperties(hmsTable);
+    try {
+      this.icebergTable = Catalogs.loadTable(conf, catalogProperties);
+    } catch (NoSuchTableException nte) {
+      // iceberg table was not yet created, no need to delete the metadata dir separately
+      return;
+    }
+    // we want to keep the data files but get rid of the metadata directory
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:276]: nit: in iceberg code we usually put an extra line after blocks

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 3]
```diff
@@ -254,6 +262,32 @@ public void commitAlterTable(org.apache.hadoop.hive.metastore.api.Table hmsTable
...
+    hmsTable.getParameters().put(InputFormatConfig.EXTERNAL_TABLE_PURGE, "FALSE");
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:277]: Why is this needed?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 4]
```diff
@@ -254,6 +262,32 @@ public void commitAlterTable(org.apache.hadoop.hive.metastore.api.Table hmsTable
     }
   }
 
+  @Override
+  public void rollbackAlterTable(org.apache.hadoop.hive.metastore.api.Table hmsTable, EnvironmentContext context)
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:266]: If this is called for any alter table op, will we be able to recognise the operation type and delete the metadata dir only in case of a true migration (and not do it for an alter table rename column, a simple alter table setprop case, etc)?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 5]
```diff
@@ -254,6 +262,32 @@ public void commitAlterTable(org.apache.hadoop.hive.metastore.api.Table hmsTable
...
+    this.catalogProperties = getCatalogProperties(hmsTable);
+    try {
+      this.icebergTable = Catalogs.loadTable(conf, catalogProperties);
+    } catch (NoSuchTableException nte) {
+      // iceberg table was not yet created, no need to delete the metadata dir separately
+      return;
+    }
+    // we want to keep the data files but get rid of the metadata directory
+    hmsTable.getParameters().put(InputFormatConfig.EXTERNAL_TABLE_PURGE, "FALSE");
+    String metadataLocation = ((BaseTable) this.icebergTable).operations().current().metadataFileLocation();
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:278]: Shouldn't we just use `CatalogUtil.dropTableData(deleteIo, deleteMetadata);`?

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java [hunk 6]
```diff
@@ -194,6 +203,88 @@ public void testScanTable() throws IOException {
     Assert.assertArrayEquals(new Object[] {"Alice", 0L}, descRows.get(2));
   }
 
+  @Test
+  public void testMigrateHiveTableToIceberg() {
+    Assume.assumeTrue("migration is only supported for hive catalog",
```
Comment (pvary) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:208]: What happens if different catalog is configured?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 7]
```diff
@@ -254,6 +262,32 @@ public void commitAlterTable(org.apache.hadoop.hive.metastore.api.Table hmsTable
...
+    try {
+      Path path = new Path(metadataLocation).getParent();
+      FileSystem fileSystem = FileSystem.get(path.toUri(), conf);
+      if (fileSystem.exists(path)) {
+        fileSystem.delete(path, true);
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:283]: Can we add some logging here that the rollback is going to happen for `tableName` and metadata under `path`?

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java [hunk 8]
```diff
@@ -955,4 +1046,44 @@ private void validateBasicStats(Table icebergTable, String dbName, String tableN
       Assert.assertEquals(summary.get(entry.getValue()), hmsParams.get(entry.getKey()));
     }
   }
+
+  private void validateMigration(String tableName, int recordCount) {
+    List<Object[]> originalResult = shell.executeStatement("SELECT * FROM " + tableName + " ORDER BY a");
```
Comment (pvary) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:1051]: Can we minimize the time spent here?
* Do we need this query, or we can expect that the writer of the test knows the expected number?
* Do we need it to be ordered?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java (hunk 3 above)
Comment (marton-bod) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:277]: Will this change actually be saved into the HMS db? If so, what if the original table had this property as true? Should we change it silently here due to the rollback?

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java [hunk 9]
```diff
@@ -955,4 +1046,44 @@ private void validateBasicStats(Table icebergTable, String dbName, String tableN
...
+    Assert.assertEquals(recordCount, originalResult.size());
+    List<Object[]> originalDescribe = shell.executeStatement("DESCRIBE FORMATTED " + tableName);
+    validateDescribeOutput(originalDescribe, fileFormat.name());
+    shell.executeStatement("ALTER TABLE " + tableName + " SET TBLPROPERTIES " +
+        "('storage_handler'='org.apache.iceberg.mr.hive.HiveIcebergStorageHandler')");
+    List<Object[]> alterResult = shell.executeStatement("SELECT * FROM " + tableName + " ORDER BY a");
```
Comment (pvary) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:1057]: Do we need this ordered?

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java [hunk 10]
```diff
@@ -955,4 +1046,44 @@ private void validateBasicStats(Table icebergTable, String dbName, String tableN
...
+      shell.executeStatement("ALTER TABLE " + tableName + " SET TBLPROPERTIES " +
+          "('storage_handler'='org.apache.iceberg.mr.hive.HiveIcebergStorageHandler')");
+      List<Object[]> alterResult = shell.executeStatement("SELECT * FROM " + tableName + " ORDER BY a");
+      Assert.assertEquals(originalResult.size(), alterResult.size());
+      List<Object[]> alterDescribe = shell.executeStatement("DESCRIBE FORMATTED " + tableName);
+      validateDescribeOutput(alterDescribe, fileFormat.name());
+    }
+  }
+
+  private void validateDescribeOutput(List<Object[]> describe, String format) {
```
Comment (pvary) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:1082]: What does this validation check?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 11]
```diff
@@ -159,11 +162,16 @@ public void commitCreateTable(org.apache.hadoop.hive.metastore.api.Table hmsTabl
 
   @Override
   public void preDropTable(org.apache.hadoop.hive.metastore.api.Table hmsTable) {
+    // do nothing
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:165]: When does this version of the hook get called (vs the one with deleteData param)?

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java [hunk 12]
```diff
@@ -955,4 +1046,44 @@ private void validateBasicStats(Table icebergTable, String dbName, String tableN
       Assert.assertEquals(summary.get(entry.getValue()), hmsParams.get(entry.getKey()));
     }
   }
+
+  private void validateMigration(String tableName, int recordCount) {
+    List<Object[]> originalResult = shell.executeStatement("SELECT * FROM " + tableName + " ORDER BY a");
+    Assert.assertEquals(recordCount, originalResult.size());
+    List<Object[]> originalDescribe = shell.executeStatement("DESCRIBE FORMATTED " + tableName);
+    validateDescribeOutput(originalDescribe, fileFormat.name());
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:1054]: I'm not sure we need this, since this is only checking hive behaviour, not iceberg

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java [hunk 13]
```diff
@@ -955,4 +1046,44 @@ private void validateBasicStats(Table icebergTable, String dbName, String tableN
...
+    shell.executeStatement("ALTER TABLE " + tableName + " SET TBLPROPERTIES " +
+        "('storage_handler'='org.apache.iceberg.mr.hive.HiveIcebergStorageHandler')");
+    List<Object[]> alterResult = shell.executeStatement("SELECT * FROM " + tableName + " ORDER BY a");
+    Assert.assertEquals(originalResult.size(), alterResult.size());
+    List<Object[]> alterDescribe = shell.executeStatement("DESCRIBE FORMATTED " + tableName);
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:1059]: Can we use `TestHiveMetastore#loadTable` instead and check the contents of the `sd`?

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java [hunk 14]
```diff
@@ -955,4 +1046,44 @@ private void validateBasicStats(Table icebergTable, String dbName, String tableN
...
+
+  private void validateMigration(String tableName, int recordCount) {
+    List<Object[]> originalResult = shell.executeStatement("SELECT * FROM " + tableName + " ORDER BY a");
+    Assert.assertEquals(recordCount, originalResult.size());
+    List<Object[]> originalDescribe = shell.executeStatement("DESCRIBE FORMATTED " + tableName);
+    validateDescribeOutput(originalDescribe, fileFormat.name());
+    shell.executeStatement("ALTER TABLE " + tableName + " SET TBLPROPERTIES " +
+        "('storage_handler'='org.apache.iceberg.mr.hive.HiveIcebergStorageHandler')");
+    List<Object[]> alterResult = shell.executeStatement("SELECT * FROM " + tableName + " ORDER BY a");
+    Assert.assertEquals(originalResult.size(), alterResult.size());
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:1058]: Would it be difficult to check the contents as well, not just the size? Just to make sure the data is all the same after the migration

File: iceberg/pom.xml [hunk 15]
```diff
@@ -31,7 +31,7 @@
         <path.to.iceberg.root>.</path.to.iceberg.root>
         <iceberg-api.version>0.11.0</iceberg-api.version>
         <kryo-shaded.version>4.0.2</kryo-shaded.version>
-        <iceberg.mockito-core.version>1.10.19</iceberg.mockito-core.version>
+        <iceberg.mockito-core.version>3.4.4</iceberg.mockito-core.version>
```
Comment (pvary) [inline iceberg/pom.xml:34]: How big is this change? Would it worth to separate to a different jira?

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java [hunk 16]
```diff
@@ -138,8 +141,32 @@ private void finalizeAlterTableWithWriteIdOp(Table table, Table oldTable, List<P
       environmentContext.putToProperties(HiveMetaHook.ALTER_TABLE_OPERATION_TYPE, desc.getType().name());
       if (partitions == null) {
         long writeId = desc.getWriteId() != null ? desc.getWriteId() : 0;
-        context.getDb().alterTable(desc.getDbTableName(), table, desc.isCascade(), environmentContext, true,
-            writeId);
+        try {
+          context.getDb().alterTable(desc.getDbTableName(), table, desc.isCascade(), environmentContext, true, writeId);
+        } catch (HiveException ex) {
+          if (environmentContext.getProperties().containsKey(HiveMetaHook.INITIALIZE_ROLLBACK_ALTER) && Boolean
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java:147]: Could `environmentContext.getProperties().getOrDefault()` work for us here?

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java [hunk 17]
```diff
@@ -138,8 +141,32 @@ private void finalizeAlterTableWithWriteIdOp(Table table, Table oldTable, List<P
...
+              .valueOf(environmentContext.getProperties().get(HiveMetaHook.INITIALIZE_ROLLBACK_ALTER))) {
+            // in case of rollback of alter table do the following:
+            // 1. drop the already altered table but keep the data files
+            // 2. recreate the original table
+            // 3. run msck repair to sync partitions on filesystem with metastore
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java:152]: This could have unexpected result, if we have extra directories in the table root, which are not added yet to the HMS. I think it would be good to have a way to disable things

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java [hunk 18]
```diff
@@ -138,8 +141,32 @@ private void finalizeAlterTableWithWriteIdOp(Table table, Table oldTable, List<P
...
+            context.getDb().dropTable(table.getDbName(), table.getTableName(), false, true, false);
+            context.getDb().createTable(oldTable);
+            Msck msck = new Msck(false, false);
+            try {
+              msck.init(Msck.getMsckConf(context.getDb().getConf()));
+              msck.updateExpressionProxy(Msck.getProxyClass(context.getDb().getConf()));
+              MsckInfo msckInfo =
+                  new MsckInfo(table.getCatalogName(), table.getDbName(), table.getTableName(), null, null, true, true,
```
Comment (marton-bod) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java:160]: I wish this `MockInfo` had a builder :) oh well

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java [hunk 19]
```diff
@@ -41,6 +39,11 @@
 import org.apache.hadoop.hive.ql.metadata.Partition;
 import org.apache.hadoop.hive.ql.metadata.Table;
 import org.apache.hadoop.hive.ql.session.SessionState;
+import org.apache.thrift.TException;
+
+import java.util.ArrayList;
```
Comment (marton-bod) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java:44]: I think spotless will complain that these imports are separated.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java (hunk 1 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:282]: Right, removed the if

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java (hunk 4 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:266]: I've marked in the EnvironmentContext that we are in the middle of a migration, so no other alter operation type remove the metadata dir.

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java (hunk 6 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:208]: It works  :) (with a bit of tweaking)

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java (hunk 5 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:278]: The `CatalogUtil.dropTableData` removes everything, including the data files.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java (hunk 3 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:277]: No, it won't :). This some leftover code from the previous implementation, which I forgot to remove.

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java (hunk 9 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:1057]: Not anymore.

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java (hunk 12 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:1054]: You are absolutely right.

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java (hunk 14 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:1058]: Added content check

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java (hunk 13 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:1059]: Sure. Thanks for pointing this out.

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java (hunk 10 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:1082]: It validates whether the contents of the SD (serde, input/output format) is changed/retained (in case of rollback). I've changed this method based on @marton-bod's suggestions.

File: iceberg/pom.xml (hunk 15 above)
Comment (lcspinter) [reply iceberg/pom.xml:34]: To bump the version in other places as well I would need to touch 7-8 modules. It's not a trivial change.

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java (hunk 19 above)
Comment (lcspinter) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java:44]: It does not :). Anyway, I reordered them.

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java (hunk 16 above)
Comment (lcspinter) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java:147]: Fixed.

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java (hunk 17 above)
Comment (lcspinter) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java:152]: I did some manual testing, with adding/removing directories to the table, and it seems `Msck#repair` is using the file system as a single source of truth. If some partitions are added/removed it updates the metastore respectively.

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java (hunk 18 above)
Comment (lcspinter) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java:160]: Well, yes :). I was thinking about this, but I would have to change to code in a dozen more places, which doesn't fit into this PR.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java (hunk 5 above)
Comment (pvary) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:278]: Shall we create a new version to remove only metadata files?

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java (hunk 17 above)
Comment (pvary) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java:152]: That's what I fear. We might end up adding partitions which are on the FS but intentionally not has not been added to the table yet. Also I seem to recall that we had issues with `MSCK REPAIR` if the file naming was not correct (wrong order of partition keys, or wrong type of values - not valid number for an int column etc...)

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java (hunk 19 above)
Comment (marton-bod) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java]: ah, yes because it's not the `hive-iceberg` module :)

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java (hunk 18 above)
Comment (marton-bod) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java:159]: Yeah, makes sense. I guess we could do it one day when we have way too much time on our hands :)

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java (hunk 17 above)
Comment (lcspinter) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java]: I've removed the drop/create/msck part, and rolled back the serde, input/output format, partition keys and table properties. An exception is also thrown to notice the user we had an issue during migration and a `msck repair table` should be issued to add the missing partition data to metastore.

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java [hunk 20]
```diff
@@ -146,22 +147,23 @@ private void finalizeAlterTableWithWriteIdOp(Table table, Table oldTable, List<P
...
+            table.getSd().setOutputFormat(oldTable.getSd().getOutputFormat());
+            table.getSd().setSerdeInfo(oldTable.getSd().getSerdeInfo());
+            table.getSd().getCols().removeAll(oldTable.getPartitionKeys());
+            table.setPartCols(oldTable.getPartitionKeys());
+
+            table.getParameters().clear();
+            table.getParameters().putAll(oldTable.getParameters());
+            context.getDb().alterTable(desc.getDbTableName(), table, desc.isCascade(), environmentContext, true, writeId);
+            throw new HiveException("Error occurred during hive table migration to iceberg. Table properties "
+                + "and serde info was reverted to it's original value. Partition info was lost during the migration "
```
Comment (marton-bod) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java:163]: nit: typo, its

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java [hunk 21]
```diff
@@ -1158,32 +1172,39 @@ private void validateBasicStats(Table icebergTable, String dbName, String tableN
...
+            "('storage_handler'='org.apache.iceberg.mr.hive.HiveIcebergStorageHandler')");
+      } catch (IllegalArgumentException e) {
+        Assert.assertTrue(e.getMessage().contains("Error occurred during hive table migration to iceberg."));
+        shell.executeStatement("MSCK REPAIR TABLE " + tableName);
+        List<Object[]> alterResult = shell.executeStatement("SELECT * FROM " + tableName + " ORDER BY a");
+        Assert.assertEquals(originalResult.size(), alterResult.size());
+        for (int i = 0; i < originalResult.size(); i++) {
+          Assert.assertTrue(Arrays.equals(originalResult.get(i), alterResult.get(i)));
+        }
+        validateSd(tableName, fileFormat.name());
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:1204]: nit: I think this sd validation check might belong logically before the msck repair command. In case the sd wasn't reverted correctly, we'd get a failure during the above select which could be harder to figure out why

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 22]
```diff
@@ -254,6 +263,35 @@ public void commitAlterTable(org.apache.hadoop.hive.metastore.api.Table hmsTable
...
+      this.catalogProperties = getCatalogProperties(hmsTable);
+      try {
+        this.icebergTable = Catalogs.loadTable(conf, catalogProperties);
+      } catch (NoSuchTableException nte) {
+        // iceberg table was not yet created, no need to delete the metadata dir separately
+        return;
+      }
+
+      // we want to keep the data files but get rid of the metadata directory
+      hmsTable.getParameters().put(InputFormatConfig.EXTERNAL_TABLE_PURGE, "FALSE");
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:282]: Maybe I'm confused, but I thought this was to be removed?

Comment (lcspinter) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:282]: I'm confused too :). This shouldn't be there.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 23]
```diff
@@ -254,6 +263,34 @@ public void commitAlterTable(org.apache.hadoop.hive.metastore.api.Table hmsTable
...
+      this.catalogProperties = getCatalogProperties(hmsTable);
+      try {
+        this.icebergTable = Catalogs.loadTable(conf, catalogProperties);
+      } catch (NoSuchTableException nte) {
+        // iceberg table was not yet created, no need to delete the metadata dir separately
+        return;
+      }
+
+      // we want to keep the data files but get rid of the metadata directory
+      String metadataLocation = ((BaseTable) this.icebergTable).operations().current().metadataFileLocation();
```
Comment (pvary) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:282]: I wanted to put a comment here, but not sure if I have done it or forget it 😄
Could we create some `CatalogUtil.dropTableMetaData(deleteIo, deleteMetadata);` based on `CatalogUtil.dropTableData(deleteIo, deleteMetadata);`?

File: iceberg/pom.xml (hunk 15 above)
Comment (pvary) [reply iceberg/pom.xml:34]: Do we want to create a different PR for it?
Would it worth the effort?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java (hunk 23 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:282]: As we discussed offline, we will leave the metadata dir delete as it is.

**Review by pvary — APPROVED** [review]
